src/erlcloud_s3.erl

```diff
@@ -1045,16 +1047,77 @@ make_link(Expire_time, BucketName, Key, Config) ->
 
 -spec get_object_url(string(), string()) -> string().
 
- get_object_url(BucketName, Key) ->
-  get_object_url(BucketName, Key, default_config()).
+get_object_url(BucketName, Key) ->
+    get_object_url(BucketName, Key, default_config()).
 
 -spec get_object_url(string(), string(), aws_config()) -> string().
 
- get_object_url(BucketName, Key, Config) ->
-  case Config#aws_config.s3_bucket_after_host of
+get_object_url(BucketName, Key, Config) ->
+    case Config#aws_config.s3_bucket_after_host of
       false -> lists:flatten([Config#aws_config.s3_scheme, BucketName, ".", Config#aws_config.s3_host, port_spec(Config), "/", Key]);
       true  -> lists:flatten([Config#aws_config.s3_scheme, Config#aws_config.s3_host, port_spec(Config), "/", BucketName, "/", Key])
-  end.
+    end.
+
+-spec get_object_url_2(string(), string(), aws_config()) -> {Host::string(), Path::string(), URL::string()}.
+get_object_url_2(BucketName, Key, Config) ->
+    _Key = case lists:prefix("/", Key) of
+               true -> Key;
+               false -> "/" ++ Key
+           end,
+
+    case Config#aws_config.s3_bucket_after_host of
+        false ->
+            Host = BucketName ++ "." ++ Config#aws_config.s3_host,
+            {Host, _Key,  lists:flatten([Config#aws_config.s3_scheme, Host, port_spec(Config), _Key])};
+        true  ->
+            Host = Config#aws_config.s3_host,
+            Path = lists:flatten(["/", BucketName, _Key]),
+            {Host, Path, lists:flatten([Config#aws_config.s3_scheme, Host, port_spec(Config), Path])}
+    end.
+
+-spec signature(aws_config(), string(), string(), string(), proplist(), proplist(), string()) -> string().
+signature(Config, Path, Date, Region, QueryParams, Headers, Payload) ->
```

Comment:
This function is just for s3, not a common signature. Maybe I can use a new function name.

Reply:
There are two ways of [signing AWS API requests](https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html#sigv4_signing-steps-overview):  using auth HTTP headers or query string. 
Existing `erlcloud_aws:sign_v4()` and `erlcloud_aws:sign_v4_headers` implement the former and return the list of signed HTTP headers.

I suggest to move all of these functions into a new  `erlcloud_aws:sign_v4_query` which implements the latter and returns query with authentication information in it. 
And let's try to make this function non S3 specific so that it can be reused for other AWS services.
As an example, EC2 also has an [Query Parameters](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Query-Requests.html#query-parameters).

It may happen that there still will be some S3 specific part though. But I feel that common query signing code can be extracted.

Reply:
Download s3 object by v4 presigned  url is very useful, for example share objects, and it is more simple than HTTP headers.
Hope update coming soon.
Thank you.